Gpg 868 migrate database

CodrinCotarlanSoftwire opened the pull request.

davidm-m commented on terraform/main.tf:
```diff
@@ -1,15 +1,62 @@
+  from_port         = local.postgres_config.port
+  to_port           = local.postgres_config.port
+  protocol          = "tcp"
+  cidr_blocks       = ["0.0.0.0/0"]
```
Do we really want to allow ingress and egress from any ip address? Is there a way to restrict this to just connections from within our vpc for instance? I don't think we need any external applications to directly connect to the database.

davidm-m commented on terraform/prod-app.tfvars:
```diff
@@ -0,0 +1,10 @@
+  instance_class = "db.t3.small"
+  identifier     = "gpg-prod-db"
+  db_name        = "gpgProdDb"
+  username       = "postgres"
+  password       = "postgres"
```
For my peace of mind, is there a future ticket to change this so we're not storing our database credentials in plain text, in the code? If so it would be nice to have a todo comment here referencing it.

CodrinCotarlanSoftwire replied:
Yes, I've moved them as environment variables in devops.

davidm-m reviewed (commented):
Mostly looks good, I've got a couple of questions I'd like answered before it's merged though.

LadunOmideyiSoftwire commented on terraform/main.tf:
```diff
@@ -1,15 +1,62 @@
+resource "aws_security_group" "allow_postgres_connection" {
+  name        = join("_", ["allow_postgres_connection", var.env])
+  description = "Allow Postgres DB traffic"
+  vpc_id      = "vpc-0e9491851e4208470"
```
vpc_id can just be a variable e.g. var.vpc_id as this will be defined in the vpc resource

So can subnet (private_subnets) but I cannot see this defined anywhere. Where are we specifying the rds subnet?

LadunOmideyiSoftwire commented on terraform/main.tf:
```diff
@@ -1,15 +1,62 @@
+    engine                      = "postgres"
+    engine_version              = "14"
+    port                        = 5432
+    backup_retention_period     = 5
```
We should make longer. Maybe 15 - 30 days?

CodrinCotarlanSoftwire replied:
Was not sure about the number. I know gov paas performs weekly backups. I'll put 15 in there then.

LadunOmideyiSoftwire commented on terraform/main.tf:
```diff
@@ -1,15 +1,62 @@
+resource "aws_db_instance" "gpg-dev-db" {
+  allocated_storage           = local.postgres_config.allocated_storage
+  engine                      = local.postgres_config.engine
+  engine_version              = local.postgres_config.engine_version
```
Does it matter that this is a newer version than used in the codebase?

CodrinCotarlanSoftwire replied:
It should make no difference. Locally I installed the latest version by default and it works without issue, so this should too.

CodrinCotarlanSoftwire commented on terraform/main.tf:
```diff
@@ -1,15 +1,62 @@
+  publicly_accessible         = local.postgres_config.publicly_accessible
+  allow_major_version_upgrade = local.postgres_config.allow_major_version_upgrade
+  multi_az                    = local.postgres_config.multi_az
+  skip_final_snapshot         = local.postgres_config.skip_final_snapshot
```
I don't think there's any reason. I couldn't find whether the current configuration skips it or not. There's no harm in _not_ skipping it. I'll Add it.

davidm-m commented on terraform/variables.tf:
```diff
@@ -24,8 +24,8 @@ variable "postgres_config" {
-    password                    = string
+    username                    = optional(string)
+    password                    = optional(string)
     port                        = optional(number)
```
What happens if these aren't provided? I assume the setup will fail, in which case they probably shouldn't be optional?

CodrinCotarlanSoftwire commented:
Continuing from our work with @LadunOmideyiSoftwire, I have, in essence:

- added the VPC module (instead of creating the resources)
- Removed the `defaults` experimental module
- split each variable in the postgres object into it's own (each with its own description and sometimes default)

CodrinCotarlanSoftwire commented on terraform/.terraform.lock.hcl:
```diff
@@ -0,0 +1,62 @@
+# This file is maintained automatically by "terraform init".
+# Manual edits may be lost in future updates.
```
Terraform documentation suggests to commit this file. See https://stackoverflow.com/questions/67963719/should-terraform-lock-hcl-be-included-in-the-gitignore-file

CodrinCotarlanSoftwire commented on terraform/postgres.tf:
```diff
@@ -0,0 +1,55 @@
+
+// The database resource
+resource "aws_db_instance" "gpg-dev-db" {
+  allocated_storage           = var.rds_config_allocated_storage
+  engine                      = var.rds_config_engine
+  engine_version              = var.rds_config_engine_version
+  instance_class              = var.rds_config_instance_class
+  identifier                  = var.rds_config_identifier
+  db_name                     = var.rds_config_db_name
+  username                    = var.POSTGRES_CONFIG_USERNAME
+  password                    = var.POSTGRES_CONFIG_PASSWORD
+  port                        = var.rds_config_port
+  backup_retention_period     = var.rds_config_backup_retention_period
+  backup_window               = var.rds_config_backup_window
+  vpc_security_group_ids      = [aws_security_group.allow_postgres_connection.id]
+  db_subnet_group_name        = module.vpc.database_subnet_group_name
+  storage_encrypted           = var.rds_config_storage_encrypted
+  publicly_accessible         = var.rds_config_publicly_accessible
+  allow_major_version_upgrade = var.rds_config_allow_major_version_upgrade
+  multi_az                    = var.rds_config_multi_az
+  skip_final_snapshot         = var.rds_config_skip_final_snapshot
```
This may be a topic for the future, but we should discuss sensible default values for each of these, and what the dev / preprod / prod values should be.

CodrinCotarlanSoftwire commented on terraform/provider.tf:
```diff
@@ -1,30 +1,27 @@
+
+  }
+
+  backend "s3" {}
```
Declared in .tfbackend files

davidm-m commented on terraform/vpc.tf:
```diff
@@ -0,0 +1,47 @@
+  cidr = "10.0.0.0/16"
+
+  azs              = slice(data.aws_availability_zones.available.zone_ids, 0, 2)
+  public_subnets   = [cidrsubnet("10.0.0.0/16", 4, 0), cidrsubnet("10.0.0.0/16", 4, 1)]
+  database_subnets = [cidrsubnet("10.0.0.0/16", 4, 2), cidrsubnet("10.0.0.0/16", 4, 3)]
```
This looks worryingly like our database subnets are a subset of our public subnets.

CodrinCotarlanSoftwire replied:
database_subnets is a subset of subnets, and is private by default.

davidm-m commented on terraform/vpc.tf:
```diff
@@ -0,0 +1,47 @@
+  enable_dns_hostnames = true
+  enable_dns_support   = true
+
+  enable_nat_gateway                     = false
+  single_nat_gateway                     = true
```
Don't these contradict each other? I don't think we'll be creating any NAT gateways with `enable_nat_gateway` set to `false`

davidm-m reviewed (changes requested):
Got a few concerns, but you guys know more than me about Terraform so feel free to argue with me on them.

LadunOmideyiSoftwire commented on terraform/variables.tf:
```diff
@@ -9,4 +9,104 @@ variable "example_s3_versioning_enabled" {
+variable "rds_config_allow_major_version_upgrade" {
+  type        = bool
+  default     = true
+  description = "Specifies if the database can update major versions e.g. 11 -> 12"
```
I agree it's risky to upgrade so we can turn this to no. I think we should be able to pop a cloudwatch alarm into the terraform config to give us notice of a major version change so that at least we are aware of upgrades.

davidm-m commented on terraform/postgres.tf:
```diff
@@ -0,0 +1,55 @@
+  to_port           = 0
+  protocol          = "-1"
+  cidr_blocks       = ["0.0.0.0/0"]
+}
```
Why do we want our db to respond anywhere? Surely it should only be talking to connections from our VPC? Or we expecting to connect directly to the database from elsewhere?

davidm-m reviewed (changes requested):
Still concerned with allowing universal egress on the db - by default we should be locking down connections (in and out) as much as possible, so if we don't need the db to connect to everywhere we shouldn't let it.

CodrinCotarlanSoftwire commented:
Since, as you mention, we do not plan on connecting to the database from anything else except the app I believe it is fine to restrict its entire traffic inside he VPC, ingress and egress.

davidm-m reviewed (approved):
Lovely stuff.
